hack/test-go.sh

```diff
@@ -1,129 +1,250 @@
+        # henceforth as packages and instead interpret them as flags to give to `go test`
+        break
+    fi
+    # an arg found before the first flag is a package
+    package_args+=" ${arg}"
+    shift
+done
+gotest_flags+=" $*"
+
+# Determine packages to test
+test_packages=
+if [[ -n "${package_args}" ]]; then
+    for package in ${package_args}; do
+        test_packages="${test_packages} ${OS_GO_PACKAGE}/${package}"
+    done
+else
+    # If no packages are given to test, we need to generate a list of all packages with unit tests
+    openshift_test_packages="$(list_test_packages_under '*')"  
+
+    kubernetes_path="Godeps/_workspace/src/k8s.io/kubernetes"
+    mandatory_kubernetes_packages="${OS_GO_PACKAGE}/${kubernetes_path}/pkg/api ${OS_GO_PACKAGE}/${kubernetes_path}/pkg/api/v1"
+
+    test_packages="${openshift_test_packages} ${mandatory_kubernetes_packages}"
+
+    if [[ -n "${test_kube}" ]]; then
+        # we need to find all of the kubernetes test suites, excluding those we directly whitelisted before, the end-to-end suite, and
+        # the go2idl tests which we currently do not support
+        optional_kubernetes_packages="$(find "${kubernetes_path}" -not \(                             \
+          \(                                                                                          \
+            -path "${kubernetes_path}/pkg/api"                                                        \
+            -o -path "${kubernetes_path}/pkg/api/v1"                                                  \
+            -o -path "${kubernetes_path}/test/e2e"                                                    \
+            -o -path "${kubernetes_path}/cmd/libs/go2idl/client-gen/testoutput/testgroup/unversioned" \
+          \) -prune                                                                                   \
+        \) -name '*_test.go' -print0 | xargs -0n1 dirname -z | sort -uz | xargs -0n1 printf "${OS_GO_PACKAGE}/%s\n")"
+
+        test_packages="${test_packages} ${optional_kubernetes_packages}"
+    fi
+fi
 
-  echo 'mode: atomic' > ${OUTPUT_COVERAGE}/profiles.out
-  find $OUTPUT_COVERAGE -name profile.out | xargs sed '/^mode: atomic$/d' >> ${OUTPUT_COVERAGE}/profiles.out
-  go tool cover "-html=${OUTPUT_COVERAGE}/profiles.out" -o "${OUTPUT_COVERAGE}/coverage.html"
+if [[ -n "${dry_run}" ]]; then
+    echo "The following base flags for \`go test\` will be used by $0:"
+    echo "go test ${gotest_flags}"
+    echo "The following packages will be tested by $0:"
+    for package in ${test_packages}; do
+        echo "${package}"
+    done
+    exit 0
+fi
 
-  rm -rf $OUTPUT_COVERAGE/$OS_GO_PACKAGE
+# Run 'go test' with the accumulated arguments and packages:
+if [[ -n "${junit_report}" ]]; then
+    # we need to generate jUnit xml
+    hack/build-go.sh tools/junitreport
+    junitreport="$(os::build::find-binary junitreport)"
+
+    if [[ -z "${junitreport}" ]]; then
```

Comment:
This is par for the course, we do it in every script that we build a tool. Why? Maybe a good question.